File: Python/Stacks & Queues/22_queue_using_2_stack.py
"""
Idea:
Insertion- transferring elements from A to B & appending the value
Deletion - transferring elements from B to A & deleting top element
"""


# User function Template for python3

# Function to push an element in queue by using 2 stacks.
def Push(x, s1, s2):
    '''
    x: value to push
    stack1: list
    stack2: list
    '''
    # Transfer all elements from s1 to s2
    while s1:
        s2.append(s1.pop())

    # Push the new element onto s2
    s2.append(x)


# Function to pop an element from queue by using 2 stacks.
def Pop(s1, s2):
    '''
    stack1: list
    stack2: list
    '''
    # Transfer all elements from s2 to s1
    while s2:
        s1.append(s2.pop())

    # Check if s1 is empty, indicating the queue is empty
    if not s1:
        return -1  # Return -1 to indicate an empty queue

    # Pop the front element from s1
    front_element = s1.pop()

    # Return the popped element
    return front_element


# {
# Driver Code Starts
# Initial Template for Python 3

# contributed by RavinderSinghPB
if __name__ == '__main__':
    test_cases = int(input())
    for cases in range(test_cases):
        qry = int(input())
        qtyp_qry = list(map(int, input().strip().split()))

        i = 0
        stack1 = []
        stack2 = []
        while i < len(qtyp_qry):
            # print(i)
            if qtyp_qry[i] == 1:
                Push(qtyp_qry[i + 1], stack1, stack2)
                # print(i)
                i += 2
            else:
                print(Pop(stack1, stack2), end=' ')
                i += 1

        print()
# } Driver Code Ends

File: C++/Stacks & Queues/22_queue_using_2_stack.cpp
/*
Idea:
Insertion- transferring elements from A to B & appending the value
Deletion - transferring elements from B to A & deleting top element
*/

//{ Driver Code Starts
#include<bits/stdc++.h>

using namespace std;

class StackQueue {
private:
    stack<int> s1;
    stack<int> s2;
public:
    void push(int B);
    int pop();

};
int main()
{

    int T;
    cin >> T;
    while (T--)
    {
        StackQueue* sq = new StackQueue();

        int Q;
        cin >> Q;
        while (Q--) {
            int QueryType = 0;
            cin >> QueryType;
            if (QueryType == 1)
            {
                int a;
                cin >> a;
                sq->push(a);
            }
            else if (QueryType == 2) {
                cout << sq->pop() << " ";

            }
        }
        cout << endl;
    }


}

// } Driver Code Ends




/* The structure of the class is
class StackQueue{
private:
    // These are STL stacks ( http://goo.gl/LxlRZQ )
    stack<int> s1;
    stack<int> s2;
public:
    void push(int);
    int pop();
}; */

//Function to push an element in queue by using 2 stacks.
// Function to push an element in the queue by using 2 stacks.
void StackQueue::push(int x)
{
    // Your Code

    // Transfer all elements from s1 to s2
    while (!s1.empty())
    {
        s2.push(s1.top());
        s1.pop();
    }

    // Push the new element onto s2
    s2.push(x);
}

// Function to pop an element from the queue by using 2 stacks.
int StackQueue::pop()
{
    // Your Code

    // Transfer all elements from s2 to s1
    while (!s2.empty())
    {
        s1.push(s2.top());
        s2.pop();
    }

    // Check if s1 is empty, indicating the queue is empty
    if (s1.empty())
        return -1; // Return -1 to indicate an empty queue

    // Pop the front element from s1
    int frontElement = s1.top();
    s1.pop();

    // Return the popped element
    return frontElement;
}
